### PR TITLE
Clarify restore-keys description

### LINKED ISSUE #783: Documentation unclear on how to specify multiple `restore-keys`
Documentation at https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows states that `restore-keys` is "An ordered list of alternative keys to use for finding the cache if no cache hit occurred for key.". However, if I supply an ordered list for `restore-keys` in a job, I get an error:

```
      - name: Cache CVE database for OWASP dependency-check
        uses: actions/cache@v3
        with:
          path: dependency-check-data
          key: "${{ runner.os }}-dependency-check-data-${{ hashFiles('**/nvdcve-1.1-modified.meta') }}"
          restore-keys:
            - "${{ runner.os }}-dependency-check-data-${{ hashFiles('**/nvdcve-1.1-modified.meta') }}"
            - "${{ runner.os }}-dependency-check-data-"
```

>[Invalid workflow file: .github/workflows/maven.yml#L37](https://github.com/*****/************/actions/runs/2096252491/workflow)
>The workflow is not valid. .github/workflows/maven.yml (Line: 37, Col: 13): A sequence was not expected

If this is not how an ordered list should be specified, a more detailed explanation in the documentation would be useful.

The example that does specify multiple `restore-keys` does it as a multi-line string, but when I try this the entire string appears to be treated as a single key?
```
      - name: Cache CVE database for OWASP dependency-check
        uses: actions/cache@v3
        with:
          path: dependency-check-data
          key: "${{ runner.os }}-dependency-check-data-${{ hashFiles('**/nvdcve-1.1-modified.meta') }}"
          restore-keys: |
            ${{ runner.os }}-dependency-check-data-${{ hashFiles('**/nvdcve-1.1-modified.meta') }}
            ${{ runner.os }}-dependency-check-data-
```

Debug output on post-step:
```
##[debug]Evaluating: format('{0}-dependency-check-data-{1}
##[debug]{2}-dependency-check-data-
##[debug]', runner.os, hashFiles('**/nvdcve-1.1-modified.meta'), runner.os)
##[debug]Evaluating format:
##[debug]..Evaluating String:
##[debug]..=> '{0}-dependency-check-data-{1}
##[debug]{2}-dependency-check-data-
##[debug]'
##[debug]..Evaluating Index:
##[debug]....Evaluating runner:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'os'
##[debug]..=> 'Linux'
##[debug]..Evaluating hashFiles:
##[debug]....Evaluating String:
##[debug]....=> '**/nvdcve-1.1-modified.meta'
##[debug]Search root directory: '/home/runner/work/********/********'
##[debug]Search pattern: '**/nvdcve-1.1-modified.meta'
##[debug]Starting process:
##[debug]  File name: '/home/runner/runners/2.289.2/externals/node16/bin/node'
##[debug]  Arguments: '"/home/runner/runners/2.289.2/bin/hashFiles"'
##[debug]  Working directory: '/home/runner/work/********/********'
##[debug]  Require exit code zero: 'False'
##[debug]  Encoding web name:  ; code page: ''
##[debug]  Force kill process on cancellation: 'False'
##[debug]  Redirected STDIN: 'False'
##[debug]  Persist current code page: 'False'
##[debug]  Keep redirected STDIN open: 'False'
##[debug]  High priority process: 'False'
##[debug]Updated oom_score_adj to 500 for PID: 2010.
##[debug]Process started with process id 2010, waiting for process exit.
##[debug]Match Pattern: **/nvdcve-1.1-modified.meta
##[debug]::debug::followSymbolicLinks 'false'
##[debug]::debug::followSymbolicLinks 'false'
##[debug]::debug::implicitDescendants 'true'
##[debug]::debug::omitBrokenSymbolicLinks 'true'
##[debug]::debug::Search path '/home/runner/work/********/********'
##[debug]/home/runner/work/********/********/dependency-check-data/nvdcache/nvdcve-1.1-modified.meta
##[debug]Found 1 files to hash.
##[debug]Hash result: 'a65b3939924c996fc1daedc8e7c7c4cdf4ea9ec5b7a3c2b495ff37333213'
##[debug]undefined
##[debug]STDOUT/STDERR stream read finished.
##[debug]STDOUT/STDERR stream read finished.
##[debug]Finished process 2010 with exit code 0, and elapsed time 00:00:00.2069719.
##[debug]..=> 'a9394c996fc1da62edc8e7c7c4cdf4ea9ec5b7a3c2b495ff37333213'
##[debug]..Evaluating Index:
##[debug]....Evaluating runner:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'os'
##[debug]..=> 'Linux'
##[debug]=> 'Linux-dependency-check-data-a65b5739924c996fc1da62edc8e7c7c4cdf4ea9ec5b7a3c2b4ff37333213
##[debug]Linux-dependency-check-data-
##[debug]'
##[debug]Result: 'Linux-dependency-check-data-a65b573939924c9fc1da62edc8e7c7c4cdf4ea9ec5b7a3c2b495ff37333213
##[debug]Linux-dependency-check-data-
##[debug]'
```

### PR DESCRIPTION
**Pull request by lucascosti**
### Why:

Closes https://github.com/actions/cache/issues/783

### What's being changed:

Clarifies the description of `restore-keys`. Whilst the previous description wasn't totally incorrect, it could understandably be misinterpreted to be referring to a yaml list instead of a multiline string containing a list of ordered keys. The new description is more explicit.

### Check off the following:

- [x] I have reviewed my changes in staging (look for "Automatically generated comment" and click **Modified** to view your latest changes).
- [x] For content changes, I have completed the [self-review checklist](https://github.com/github/docs/blob/main/contributing/self-review.md#self-review).

### CODE REVIEW

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md [hunk 1]
```diff
@@ -102,7 +102,7 @@ For more information, see [`actions/cache`](https://github.com/actions/cache).
           ~/.gradle/wrapper
     ```
   - With `v1` of the `cache` action, only a single path is supported and it must be a directory. You cannot cache a single file.
-- `restore-keys`: **Optional** An ordered list of alternative keys to use for finding the cache if no cache hit occurred for `key`.
+- `restore-keys`: **Optional** A multiline string containing alternative restore keys, with each restore key being on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache.
```
Comment (vgrl) [inline content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]: Optional: Perhaps s/being/placed?
Also, are "sequentially" + "in the order provided" repeating the the intent?

**Review by vgrl — DISMISSED** [review]
Looks good. I left one optional comment. 🚀 ❇️

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md (hunk 1 above)
Comment (lucascosti) [reply content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]: > Optional: Perhaps s/being/placed?

Yep, I like that!

```suggestion
- `restore-keys`: **Optional** A multiline string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache.
```

>  Also, are "sequentially" + "in the order provided" repeating the the intent?

Not really; "sequentially" meaning that it searches for them one after the other, and "in the order provided" being the order in which they're checked. I'm open to changing it if you can think of a better wording.

Comment (vgrl) [reply content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md]: Fair enough. 🤷 I can read and interpret it both ways, so I'll leave the final decision to change (or not!) up to you.

**Comment by Benghie12345** [comment]
9679939893

នៅ​ថ្ងៃទី សុក្រ 8 មេសា 2022, 09:30 Vanessa ***@***.***>
បាន​សរសេរ​ថា៖

**Comment by vsvipul** [comment]
@lucascosti The thing here is that there can be both cases - a multiline string with multiple keys in `restore-keys` and also, a single key without the | character. So there can be these 2 cases-

```
        restore-keys: |
          "${{ runner.os }}-${{ github.sha }}-key"
          "test-key"
```

```
        restore-keys: "${{ runner.os }}-${{ github.sha }}-key"
```

So the user should be able to infer this from the docs. So, we can consider adding these code snippets as well.

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md [hunk 2]
```diff
@@ -102,7 +102,7 @@ For more information, see [`actions/cache`](https://github.com/actions/cache).
           ~/.gradle/wrapper
     ```
   - With `v1` of the `cache` action, only a single path is supported and it must be a directory. You cannot cache a single file.
-- `restore-keys`: **Optional** An ordered list of alternative keys to use for finding the cache if no cache hit occurred for `key`.
+- `restore-keys`: **Optional** A multiline string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache.
```
Comment (lucascosti) [inline content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]:
```suggestion
- `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache.
```

**Comment by lucascosti** [comment]
> @lucascosti The thing here is that there can be both cases

Thanks @vsvipul. I've removed 'multiline' from the description to cater for both situations.

> So, we can consider adding such code snippets as well.

The example in the next section serves as the example for the attributes, so I think this fix should suffice for the time-being; I will add something in the upcoming plan for the review of the caching docs to have a look at parameter-specific examples as well.

**Comment by vsvipul** [comment]
@lucascosti  Looks good. Can we add a line that tells the user to use `|` in case multiple restore keys are used? That is also confusing right now for the user.

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md [hunk 3]
```diff
@@ -102,7 +102,7 @@ For more information, see [`actions/cache`](https://github.com/actions/cache).
           ~/.gradle/wrapper
     ```
   - With `v1` of the `cache` action, only a single path is supported and it must be a directory. You cannot cache a single file.
-- `restore-keys`: **Optional** An ordered list of alternative keys to use for finding the cache if no cache hit occurred for `key`.
+- `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache.
```
Comment (lucascosti) [inline content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]: @vsvipul ok, I think we'll just end up adding a brief example ahead of the review plan. 

````suggestion
- `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache. For example:

  {% raw %}
  ```yaml
  restore-keys: |
    npm-foobar-${{ hashFiles('package-lock.json') }}
    npm-foobar-
    npm-
  ```
  {% endraw %}
````

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md [hunk 4]
```diff
@@ -102,7 +102,16 @@ For more information, see [`actions/cache`](https://github.com/actions/cache).
           ~/.gradle/wrapper
     ```
   - With `v1` of the `cache` action, only a single path is supported and it must be a directory. You cannot cache a single file.
-- `restore-keys`: **Optional** An ordered list of alternative keys to use for finding the cache if no cache hit occurred for `key`.
+- `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`, these restore keys are used sequentially in the order provided to find and restore a cache. For example:
```
Comment (martin389) [inline content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]:
```suggestion
- `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`,  an alternative cache is identified by sequentially applying the restore keys in the order provided. For example:
```

**Review by martin389 — APPROVED** [review]
LGTM, added an optional suggestion ✨

File: content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md (hunk 4 above)
Comment (lmlkxx21) [reply content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]:
```
- - [ ] > ~~
~~
```

Comment (lucascosti) [reply content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]: Hmmm, I don't think that 'applying' is the right word here, but thanks for the suggestion!

Comment (lmlkxx21) [reply content/actions/using-workflows/caching-dependencies-to-speed-up-workflows.md:105]: > ```suggestion
> - `restore-keys`: **Optional** A string containing alternative restore keys, with each restore key placed on a new line. If no cache hit occurred for `key`,  an alternative cache is identified by sequentially applying the restore keys in the order provided. For example:
> ```

**Comment by vsvipul** [comment]
Thank you @lucascosti and @martin389 . Change looks good.

**Pull request merged**